dll/win32/shell32/CCopyAsPathMenu.cpp

```diff
@@ -0,0 +1,228 @@
+                if (SUCCEEDED(hr))
+                {
+                    STRRET strret;
+                    hr = sf->GetDisplayNameOf(child, SHGDN_FORPARSING, &strret);
+                    if (SUCCEEDED(hr))
+                    {
+                        CComHeapPtr<WCHAR> path;
+                        hr = StrRetToStrW(&strret, child, &path);
+                        if (SUCCEEDED(hr))
+                        {
+                            AppendToPathList(paths, path, i);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    else
+    {
+        FORMATETC fmte = { CF_HDROP, NULL, DVASPECT_CONTENT, -1, TYMED_HGLOBAL };
+        STGMEDIUM stgm;
+        hr = pdto->GetData(&fmte, &stgm);
+        if (SUCCEEDED(hr))
+        {
+            for (i = 0, count = DragQueryFileW((HDROP)stgm.hGlobal, -1, NULL, 0); i < count && SUCCEEDED(hr); ++i)
+            {
+                WCHAR path[MAX_PATH];
+                if (DragQueryFileW((HDROP)stgm.hGlobal, i, path, _countof(path)))
+                {
+                    AppendToPathList(paths, path, i);
+                }
+            }
+            ReleaseStgMedium(&stgm);
+        }
+    }
+#endif
+
+    if (SUCCEEDED(hr))
+    {
+        DWORD err = SetClipboardFromString(paths);
+        hr = HRESULT_FROM_WIN32(err);
+    }
+
+    return hr;
+}
+
+STDMETHODIMP
+CCopyAsPathMenu::QueryContextMenu(HMENU hMenu, UINT indexMenu, UINT idCmdFirst, UINT idCmdLast, UINT uFlags)
+{
+    MENUITEMINFOW mii;
+    UINT Count = 0;
+
+    TRACE("CCopyAsPathMenu::QueryContextMenu(%p %p %u %u %u %u)\n", this,
+          hMenu, indexMenu, idCmdFirst, idCmdLast, uFlags);
+
+    // Detect SHIFT key press
+    // FIXME: GetKeyState shouldn't be called here, clients need to set the CMF_EXTENDEDVERBS flag.
+    if (!(uFlags & CMF_EXTENDEDVERBS) && !(GetKeyState(VK_SHIFT) & 0x8000))
+        return MAKE_HRESULT(SEVERITY_SUCCESS, 0, idCmdFirst);
+
+    if (uFlags & (CMF_NOVERBS | CMF_VERBSONLY))
+        return MAKE_HRESULT(SEVERITY_SUCCESS, 0, idCmdFirst);
+
+    m_idCmdFirst = idCmdFirst;
+
+    // Insert "Copy as path"
+    CStringW strText(MAKEINTRESOURCEW(IDS_COPYASPATHMENU));
+    ZeroMemory(&mii, sizeof(mii));
+    mii.cbSize = sizeof(mii);
+    mii.fMask = MIIM_ID | MIIM_TYPE;
+    mii.fType = MFT_STRING;
+    mii.wID = m_idCmdFirst;
+    mii.dwTypeData = strText.GetBuffer();
+    if (InsertMenuItemW(hMenu, indexMenu, TRUE, &mii))
+    {
+        m_idCmdAction = m_idCmdFirst;
+        ++indexMenu;
+        ++Count;
+    }
+
+    return MAKE_HRESULT(SEVERITY_SUCCESS, 0, idCmdFirst + Count);
+}
+
+STDMETHODIMP
+CCopyAsPathMenu::InvokeCommand(LPCMINVOKECOMMANDINFO lpcmi)
+{
+    HRESULT hr = E_FAIL;
+
+    TRACE("CCopyAsPathMenu::InvokeCommand(%p %p)\n", this, lpcmi);
+
+    if (IS_INTRESOURCE(lpcmi->lpVerb))
+    {
+        if (m_idCmdFirst + LOWORD(lpcmi->lpVerb) == m_idCmdAction)
+            hr = DoCopyAsPath(m_pDataObject);
```

Comment:
Note: Alternatively, you could directly do a `return DoCopyAsPath(...)` here, and a `return E_FAIL;` below. (This would eliminate the need for an explicit `hr` value. Up to you.)

Reply:
Done.